crates/router/src/core/payments/operations/payment_confirm.rs

```
@@ -699,3 +721,79 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve> ValidateRequest<F, api::Paymen
+                        if request_surcharge_details.surcharge_amount != attempt_surcharge_amount
+                            || request_surcharge_details.tax_amount != payment_attempt.tax_amount
+                        {
+                            return invalid_surcharge_details_error;
+                        }
+                    } else {
+                        // is not sent in payment create
+                        // verify that any calculated surcharge sent in session flow is same as the one sent in confirm
+                        match get_individual_surcharge_detail_from_redis(
+                            state,
+                            &payment_method_type.into(),
+                            &payment_method_type,
+                            None,
+                            &payment_attempt.attempt_id,
+                        )
+                        .await
+                        {
+                            Ok(surcharge_details) => {
+                                if surcharge_details
+                                    .is_request_surcharge_matching(request_surcharge_details)
+                                {
+                                    return Ok(());
+                                } else {
+                                    return invalid_surcharge_details_error;
+                                }
+                            }
+                            Err(err) if err.current_context() == &RedisError::NotFound => {
+                                if request_surcharge_details.surcharge_amount == 0
+                                    && request_surcharge_details.tax_amount.unwrap_or(0) == 0
+                                {
+                                    return Ok(());
+                                } else {
+                                    return invalid_surcharge_details_error;
+                                }
+                            }
+                            Err(_) => {
```

don't discard the error, lift the into InternalServerError.